### PR TITLE
New utilities genBundleSchema and genGraphQl to help when creating transfomers.

### PR DESCRIPTION
**Pull request by drorva**
### genBundleSchema.js 
 Parse the openAPI document and generate from it a list of bundles and for each bundles, the list of fields.                                                    
 That should help us when generating new transformers    

### genGraphQl.js
Generate the node files from the graphQl output in the pages.json file so we can compare them to the node files that we create.

## Testing done
Ran locally multiple times.

### CODE REVIEW

**Comment by drorva** [comment]
eslint complained about my use of console, so I had to add comments about the console lines
/* eslint-disable no-console */
because I wasn't sure what logging method to use instead.

File: script/cms/genBundleSchema.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    fs.writeFileSync('fullSchema.json', prettyOutJson);
+  }
+
+  /**
+   * Find the names of the bundles in the schema
+   * info into a file schema.json
+   */
+  loadBundles() {
+    let ii = 0;
+    _.map(this.schema, (value, key) => {
```
Comment (cvalarida) [inline script/cms/genBundleSchema.js:57]: Can this be accomplished with `this.schema.forEach()`? (Or perhaps `Object.entries(this.schema).forEach()` if `this.schema` is an object.)

File: script/cms/genBundleSchema.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      const bundleType = strs[0];
+      const bundle = strs[1];
+      // This is the convention for the bundle name
+      const bundleName = `${bundleType}-${bundle}`;
+
+      for (let jj = 0; jj < ignoredBundleTypes.length; jj++) {
+        if (bundleType === ignoredBundleTypes[jj]) {
+          return;
+        }
+      }
```
Comment (cvalarida) [inline script/cms/genBundleSchema.js:68]:
```suggestion
      if (ignoredBundleTypes.includes(bundleType)) return;
```

File: script/cms/genBundleSchema.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    });
+
+    const relationships =
+      bundleSchema.properties.data.properties.relationships.properties;
+    _.map(relationships, (value, key) => {
+      if (key.startsWith('field_')) {
+        fieldObj[key] = value;
+      }
+    });
```
Comment (cvalarida) [inline script/cms/genBundleSchema.js:96]: This should probably be a [`.reduce()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce).

```js
const attributes = bundleSchema.properties.data.properties.attributes.properties;
const relationships = bundleSchema.properties.data.properties.relationships.properties

const fieldObj = attributes.concat(relationships).reduce( ... );
```

File: script/cms/genBundleSchema.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    const fieldObj = {};
+
+    // Get the fields in this bundle
+    const fields =
+      bundleSchema.properties.data.properties.attributes.properties;
+    _.map(fields, (value, key) => {
+      if (key.startsWith('field_')) {
+        fieldObj[key] = {
+          type: value.type,
+        };
```
Comment (cvalarida) [inline script/cms/genBundleSchema.js:86]: Why pull out the type and ignore everything else?

File: script/cms/genGraphQl.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+/*
+* Save a single node to file
+*/
+
+function saveNode(node) {
+  // Create the dir if it doesn't exist
+  if (!fs.existsSync(graphqlDir)) {
+    fs.mkdirSync(graphqlDir);
+  }
```
Comment (cvalarida) [inline script/cms/genGraphQl.js:22]: Alternatively, we can use [`ensureDirSync`](https://github.com/jprichardson/node-fs-extra/blob/master/docs/ensureDir-sync.md) if we use `fs-extra`.
```suggestion
  fs.ensureDirSync(graphqlDir);
```
However, if we use [`fs.outputJsonSync()`](https://github.com/jprichardson/node-fs-extra/blob/master/docs/outputJson-sync.md), we don't need this bit at all, as it'll ensure the directories are created.

File: script/cms/genGraphQl.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  const prettyOutJson = JSON.stringify(node, null, 2);
+  if (node.entityId) {
+    const id = `${node.entityId}`;
+    /* eslint-disable no-console */
+    console.log(id);
+
+    // File name starts with the node id for easy file completion
+    const fileName = `${graphqlDir}/${id}.json`;
+    fs.writeFileSync(fileName, prettyOutJson);
```
Comment (cvalarida) [inline script/cms/genGraphQl.js:32]:
```suggestion
    fs.outputJsonSync(fileName, node, { spaces: 2 });
```

File: script/cms/genGraphQl.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
+/**
+ * Generate the node files from the grapQl output in the pages.json file
+ * so we can compare them to the node files that we generate
+ */
+
+const _ = require('lodash');
+const fs = require('fs');
```
Comment (cvalarida) [inline script/cms/genGraphQl.js:7]:
```suggestion
const fs = require('fs-extra');
```

File: script/cms/genGraphQl.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  }
+}
+
+/**
+ * Load the graphQL generated file
+ */
+
+function load() {
+  const stringData = fs.readFileSync(conf.graphqlFile);
+  const data = JSON.parse(stringData);
```
Comment (cvalarida) [inline script/cms/genGraphQl.js:45]:
```suggestion
  const data = fs.readJsonSync(conf.graphqlFile);
```

File: script/cms/genGraphQl.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+  // This is where in the graph the nodes are
+  const entities = data.data.nodeQuery.entities;
+
+  // Traverse the nodes
+  _.map(entities, (value, key) => {
+    saveNode(value);
+  });
```
Comment (cvalarida) [inline script/cms/genGraphQl.js:53]: Assuming this is an array (it should be, right?):
```suggestion
  entities.forEach(value => {
    saveNode(value);
  });
```

**Review by cvalarida — APPROVED** [review]
I didn't test this, but it looks generally okay. There's a lot we can do to clean it up, though.

File: script/cms/genGraphQl.js (hunk 9 above)
Comment (jhonnyoddball) [reply script/cms/genGraphQl.js:50]: This is an object

**Comment by jhonnyoddball** [comment]
## genGraphQL.js

Modifications were made to generate entity nodes from GraphQL and from transformed output schemas

## Screenshots

Screen Shot 2020-08-18 at 4 11 40 PM

## Commands
- Create entity nodes into `nodes` folder from GraphQL
`node script/cms/genGraphQl.js`

- Create entity nodes into `nodesTransformed` folder from transformed output schemas
`node script/cms/genGraphQl.js --transformers`

## Testing
Locally

File: src/site/stages/build/drupal/metalsmith-drupal.js [hunk 10]
```diff
@@ -222,6 +224,9 @@ async function loadDrupal(buildOptions) {
     ? await getContentFromExport(buildOptions)
     : await getContentViaGraphQL(buildOptions);
 
+  // Dynamic GraphQL from CMS build
+  fs.outputJsonSync(CMS_EXPORT_CACHE_FILENAME, drupalPages);
```
Comment (jhonnyoddball) [inline src/site/stages/build/drupal/metalsmith-drupal.js:228]: Stores all transformed data into a file

**Pull request merged**